Add Ring config flow

balloob opened the pull request:
## Breaking Change

Ring is now configurable via a config entry. This means it will no longer store the username and password and you're able to configure accounts that use 2FA. It's no longer possible to set scan interval, monitored conditions or ffmpeg arguments. The defaults are used.

## Description:
This adds config flow support to Ring.

Benefits:
 - No longer need to store username/password in config. Once logged in we just store the tokens.
 - Support two-factor auth

CC @tchellomello 

resolves #25088

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

balloob committed:
Add Ring config flow

MartinHjelmare commented on homeassistant/components/ring/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+            return self.async_abort(reason="already_configured")
+
+        await self.async_set_unique_id(user_input["username"])
+        self._abort_if_unique_id_configured()
```
When will this have an effect? Only one entry is allowed, right?

balloob replied:
Oh not necessary anymore because I cancel when we already have configured entries

MartinHjelmare commented on tests/components/ring/test_config_flow.py:
```diff
@@ -0,0 +1,59 @@
+    )
+
+    with patch(
+        "homeassistant.components.ring.config_flow.validate_input",
```
Patch the `Ring` class instead to raise `AccessDeniedError` on instantiation.

tchellomello commented:
@balloob this is awesome!! I'm going to test it tonight!

MartinHjelmare reviewed (changes requested):
Also remove the stale `CONF_SCAN_INTERVAL` option from the config schema.

tchellomello commented:
@balloob I was able to get it working. 

Basically one a fresh new installation, I went to `Configure` -> `Integrations`  -> Ring. 

Added my username and password there and the component got added correctly, however, no image was displayed so far. 

Then I modified the `configuration.yaml`  including the `ring` component to the `camera`  platform, I was able to see the video. 

![image](https://user-images.githubusercontent.com/809840/71960080-23aa4800-31c2-11ea-81a4-35380f9bcaed.png)


Tomorrow I'll enable the 2FA to test as well. So far, so good :)  :+1:

balloob commented:
Yeah I realized I need to forward the config entry. Haven't done that yet. I'll clean it up. Still going to add it to the beta

balloob committed:
Address comments + migrate platforms to config entry

balloob commented:
Addressed comments and converted the platforms to use config entries too. Now it's able to delete the entry without restart too. 

Just adding device info would make this 🔥 but I don't have any Ring devices to work with. Just using an empty account. That's for someone else.

balloob committed:
Migrate camera too

dshokouhi commented:
This should help to close #25088 :)

MartinHjelmare commented on homeassistant/components/ring/camera.py:
```diff
@@ -20,69 +18,37 @@
-    # show notification for all cameras without an active subscription
-    if cams_no_plan:
-        cameras = str(", ".join([camera.name for camera in cams_no_plan]))
+            cams.append(RingCam(hass, camera))
```
There's a side effect here. Getting the recording url makes a request.

MartinHjelmare commented on homeassistant/components/ring/camera.py:
```diff
@@ -20,69 +18,37 @@
         self._ffmpeg = hass.data[DATA_FFMPEG]
-        self._ffmpeg_arguments = device_info.get(CONF_FFMPEG_ARGUMENTS)
         self._last_video_id = self._camera.last_recording_id
         self._video_url = self._camera.recording_url(self._last_video_id)
```
This does a request.

MartinHjelmare commented on homeassistant/components/ring/__init__.py:
```diff
@@ -92,6 +132,36 @@ def hub_refresh():
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
```
There are dispatcher connections in the platforms that we should disconnect.

balloob committed:
Address comments

balloob committed:
Fix order config flows

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented on homeassistant/components/ring/__init__.py:
```diff
@@ -45,25 +51,37 @@
 
 def setup(hass, config):
```
`async_setup`

balloob committed:
setup -> async_setup

robfish1956 commented:
I tried this using docker image: homeassistant/home-assistant:0.104.0b1
and the Ring integration is available but when I submit my username and password I get
"Unexpected Error"
I have not enable 2FA

Scope666 commented:
I also have the EXACT same error as @robfish1956

balloob commented:
The ring integration has been blocked again because the library was not fixed properly last time, still causing unnecessary token refreshes on EACH request. That needs to be fixed and this will work again.
